### PR TITLE
Initial DataManager

### PR DESCRIPTION
**Pull request by Pessimistress**
[RFC](https://github.com/visgl/deck.gl/blob/master/dev-docs/RFCs/proposals/data-manager-rfc.md)

Achieved in this PR:
- Ability to register a data resource and trigger updates on the Deck level
- Multiple layers that share the same url now only load the data once

Planned for future PRs:
- Ability to share WebGL Buffer among multiple consumers of a binary
- Ability to share WebGL Texture among multiple consumers of an image

#### Change List
- `DataManager` class
- Layers by default resolve async props via shared dataManager instance
- Experimental `deck._addResources` and `deck._removeResources` methods
- Tests

### CODE REVIEW

File: modules/core/src/lib/data/data-manager.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
+/* global requestAnimationFrame */
+import DataResource from './data-resource';
```
Comment (ibgreen) [inline modules/core/src/lib/data/data-manager.js:2]: Probably too early to settle naming. It is pretty good, but a bit inconsistent. A control question could be what we will we be calling these things in the docs? "data", "resources", "data resources", "data sources" etc.

Possibilities 
- `ResourceManager` and `Resource`
- `DataManager` and `Data`
- `DataResourceManager` and `DataResource`? And all them "data resources" in the docs?

File: modules/core/src/lib/layer.js [hunk 2]
```diff
@@ -59,7 +59,28 @@ const defaultProps = {
...
+          dataManager.add(url, url, {persistent: false});
+        }
+        const resource = dataManager.subscribe(
+          url,
+          data => layer.internalState.reloadAsyncProp(propName, data),
+          layer.id,
+          propName
+        );
+        if (resource) {
+          return resource;
```
Comment (ibgreen) [inline modules/core/src/lib/layer.js:78]: What is the type of a resource? Isn't that a wrapper? Don't we want to return some data value or promise?

File: modules/core/src/lib/data/data-manager.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    }
+    if (resource) {
+      consumer[requestId] = request;
+      request.onChange = onChange;
+      request.resource = resource;
+      resource.subscribe(request);
+    }
+  }
+
+  _prune() {
```
Comment (ibgreen) [inline modules/core/src/lib/data/data-manager.js:118]: I don't understand why deleting them in a timer improves perf? Maybe add a comment

File: test/modules/core/lib/deck.spec.js [hunk 4]
```diff
@@ -167,3 +168,75 @@ test('Deck#auto view state', t => {
...
+    height: 1,
+    // This is required because the jsdom canvas does not have client width/height
+    autoResizeDrawingBuffer: gl.canvas.clientWidth > 0,
+
+    viewState: {
+      longitude: 0,
+      latitude: 0,
+      zoom: 0
+    },
+
```
Comment (ibgreen) [inline test/modules/core/lib/deck.spec.js:201]: Are we going to have a `Deck.resources` prop so we can specify these declaratively.

Comment (Pessimistress) [reply test/modules/core/lib/deck.spec.js:201]: Yes eventually

**Pull request merged**